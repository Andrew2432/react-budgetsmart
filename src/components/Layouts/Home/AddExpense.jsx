import React, { useState, useContext, Fragment, useEffect } from 'react';
import ExpenseContext from '../../../context/Expense/expenseContext';
import ToastContext from '../../../context/Toasts/ToastContext';

import { CardPanel, Row, Col, TextInput } from 'react-materialize';
import BtnAddExpense from '../Buttons/BtnAddExpense';
import BtnUpdateExpense from '../Buttons/BtnUpdateExpense';
import BtnBack from '../Buttons/BtnBack';

const AddExpense = () => {
  const expenseContext = useContext(ExpenseContext);
  const toastContext = useContext(ToastContext);

  const {
    addExpense,
    mode,
    currentExpense,
    updateExpense,
    setHomeState,
    fetchDataFromStorage,
  } = expenseContext;

  const { setToast } = toastContext;

  const [expense, setExpense] = useState('');
  const [cost, setCost] = useState('');

  const handleExpense = (e) => setExpense(e.target.value);
  const handleCost = (e) => setCost(e.target.value);

  useEffect(() => {
    if (mode === 'edit') setFields();
    // eslint-disable-next-line
  }, [mode]);

  useEffect(() => {
    fetchDataFromStorage();
    // eslint-disable-next-line
  }, []);

  const handleSubmit = (e) => {
    e.preventDefault();
    if (expense.trim() === '' || !parseInt(cost)) {
      setToast({ type: 'error', text: 'Please enter correct expense.' });
    } else {
      addExpense(expense, parseInt(cost));
      clearFields();
      setToast({ type: 'success', text: 'Added successfully' });
    }
  };

  const handleUpdate = () => {
    updateExpense(expense, parseInt(cost));
    clearFields();
  };

  const handleBack = () => {
    clearFields();
    setHomeState();
  };

  const setFields = () => {
    setExpense(currentExpense.name);
    setCost(currentExpense.cost.toString());
  };

  const clearFields = () => {
    setExpense('');
    setCost('');
  };

  return (
    <CardPanel className="deep-purple white-text">
      <h4>Add A New Expense</h4>
      <Row>
        <Col s={8} m={8} l={8}>
          <TextInput
            id="expense"
            label="Expense"
            s={12}
            m={12}
            l={12}
            value={expense}
            onChange={handleExpense}
          />
        </Col>
        <Col s={4} m={4} l={4}>
          <TextInput
            id="cost"
            label="Cost"
            s={12}
            m={12}
            l={12}
            value={cost}
            onChange={handleCost}
          />
        </Col>
      </Row>
      <Row>
        {mode === 'add' && <BtnAddExpense onClick={handleSubmit} />}
        {mode === 'edit' && (
          <Fragment>
            <BtnUpdateExpense onClick={handleUpdate} />
            <BtnBack onClick={handleBack} />
          </Fragment>
        )}
      </Row>
    </CardPanel>
  );
};

export default AddExpense;
