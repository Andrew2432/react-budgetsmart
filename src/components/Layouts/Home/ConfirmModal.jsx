import React from 'react';
import PropTypes from 'prop-types';
import { Modal, Button } from 'react-materialize';

const ConfirmModal = ({ onClick, trigger, message, id, targetid }) => {
  return (
    <Modal
      className="black-text"
      targetid={targetid ? targetid : null}
      actions={[
        <Button
          flat
          modal="close"
          node="button"
          waves="red"
          className="red white-text"
          onClick={onClick(targetid)}
        >
          Yes
        </Button>,
        <Button flat modal="close" node="button" waves="green">
          Close
        </Button>,
      ]}
      bottomSheet={false}
      fixedFooter={false}
      header="Confirm Delete?"
      id={id}
      open={false}
      options={{
        dismissible: true,
        endingTop: '10%',
        inDuration: 250,
        onCloseEnd: null,
        onCloseStart: null,
        onOpenEnd: null,
        onOpenStart: null,
        opacity: 0.5,
        outDuration: 250,
        preventScrolling: true,
        startingTop: '4%',
      }}
      trigger={trigger ? trigger : null}
    >
      <p>{message}</p>
    </Modal>
  );
};

ConfirmModal.propTypes = {
  onClick: PropTypes.func.isRequired,
  trigger: PropTypes.object,
  message: PropTypes.string.isRequired,
  id: PropTypes.string,
};

export default ConfirmModal;
