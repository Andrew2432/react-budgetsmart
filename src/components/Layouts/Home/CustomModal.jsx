import React, { Component } from 'react';
import M from 'materialize-css';

class CustomModal extends Component {
  componentDidMount() {
    const options = {
      onOpenStart: () => {
        console.log('Open Start');
      },
      onOpenEnd: () => {
        console.log('Open End');
      },
      onCloseStart: () => {
        console.log('Close Start');
      },
      onCloseEnd: () => {
        console.log('Close End');
      },
      inDuration: 250,
      outDuration: 250,
      opacity: 0.5,
      dismissible: false,
      startingTop: '4%',
      endingTop: '10%',
    };
    M.Modal.init(this.Modal, options);
  }
}
export default CustomModal;
