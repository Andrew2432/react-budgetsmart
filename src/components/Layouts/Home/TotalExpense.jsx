import React, { useContext } from 'react';
import ExpenseContext from '../../../context/Expense/expenseContext';
import { CardPanel } from 'react-materialize';

const TotalExpense = () => {
  const expenseContext = useContext(ExpenseContext);
  const { totalExpense } = expenseContext;

  return (
    <CardPanel
      className="deep-orange darken-1 white-text"
      style={{ textAlign: 'center' }}
    >
      <h4>Total expense</h4>
      <h4>Rs. {totalExpense}</h4>
    </CardPanel>
  );
};

export default TotalExpense;
