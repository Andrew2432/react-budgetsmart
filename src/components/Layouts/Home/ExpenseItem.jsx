import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { CollectionItem, Icon } from 'react-materialize';
import ExpenseContext from '../../../context/Expense/expenseContext';
import ConfirmModal from './ConfirmModal';

const ExpenseItem = (props) => {
  const expenseContext = useContext(ExpenseContext);
  const { editExpense, deleteExpense } = expenseContext;
  const { id, name, cost, createdAt } = props;

  const findID = (e) =>
    e.target.parentElement.parentElement.attributes.id.nodeValue;

  const handleEdit = (e) => {
    e.preventDefault();
    editExpense(findID(e));
  };

  const handleDelete = (id) => {
    // e.preventDefault();
    console.log(id);
    deleteExpense(id);
  };

  return (
    <CollectionItem id={id} data-key={id}>
      <b>{name}</b> | <em>Rs. {cost}</em> <br />
      <b>{createdAt}</b>
      <a className="secondary-content modal-trigger" href="#modal-1">
        <ConfirmModal
          id="modal-1"
          message="Are you sure you want to delete?"
          onClick={deleteExpense}
          targetid={id}
        />
        <Icon className="red-text">delete</Icon>
      </a>
      <a className="secondary-content" href="!#" onClick={handleEdit}>
        <Icon className="orange-text">mode_edit</Icon>
      </a>
    </CollectionItem>
  );
};

ExpenseItem.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  cost: PropTypes.number.isRequired,
  createdAt: PropTypes.string.isRequired,
};

export default ExpenseItem;
