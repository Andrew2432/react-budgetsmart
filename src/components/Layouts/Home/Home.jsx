import React, { Fragment } from 'react';
import AddExpense from './AddExpense';
import TotalExpense from './TotalExpense';

const Home = () => {
  return (
    <Fragment>
      <AddExpense />
      <TotalExpense />
    </Fragment>
  );
};

export default Home;
