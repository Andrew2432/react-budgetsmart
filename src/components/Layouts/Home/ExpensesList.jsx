import React, { useContext, useEffect } from 'react';
import {
  CardPanel,
  Collection,
  CollectionItem,
  Row,
  Col,
  Button,
} from 'react-materialize';
import ExpenseItem from './ExpenseItem';
import ConfirmModal from '../Utils/ConfirmModal';
import ExpenseContext from '../../../context/Expense/expenseContext';
import ToastContext from '../../../context/Toasts/ToastContext';

const ExpensesList = () => {
  const expenseContext = useContext(ExpenseContext);
  const toastContext = useContext(ToastContext);

  const { setToast } = toastContext;
  const {
    expenses,
    clearAllExpense,
    removeToastMode,
    toastMode,
  } = expenseContext;

  useEffect(() => {
    switch (toastMode) {
      case 'edit':
        setToast({ type: 'success', text: 'Updated successfully' });
        break;
      case 'delete':
        setToast({ type: 'success', text: 'Deleted successfully' });
        break;
      case 'clear':
        setToast({ type: 'success', text: 'Cleared successfully' });
        break;
      default:
        break;
    }
    removeToastMode();
    // eslint-disable-next-line
  }, [toastMode]);

  const createExpenseItem = (expense) => (
    <ExpenseItem {...expense} key={expense.id} />
  );

  return (
    <CardPanel className="indigo accent-2">
      <h4>Your Expenses</h4>
      <Collection className="black-text">
        {expenses.length > 0 ? (
          expenses.map(createExpenseItem)
        ) : (
          <CollectionItem>No expenses added. Enter an expense</CollectionItem>
        )}
      </Collection>
      {expenses.length > 0 && (
        <Row>
          <Col>
            <ConfirmModal
              onClick={clearAllExpense}
              trigger={
                <Button className="purple darken-2">Clear Expenses</Button>
              }
              message="Are you sure you want to delete all expenses?"
            />
          </Col>
        </Row>
      )}
    </CardPanel>
  );
};

export default ExpensesList;
