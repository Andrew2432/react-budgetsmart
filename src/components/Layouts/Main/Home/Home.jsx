import React, { Fragment, useContext } from 'react';
import AddExpense from '../AddExpense';
import TotalExpense from '../TotalExpense';
import ExpensesList from '../ExpensesList';
import CustomToast from '../../Utils/CustomToast';
import ToastContext from '../../../../context/Toasts/ToastContext';

const Home = () => {
  const toastContext = useContext(ToastContext);
  const { messages } = toastContext;

  /**
   * Maintain app level state for rendering toasts
   */
  return (
    <Fragment>
      <AddExpense />
      <TotalExpense />
      <ExpensesList />
      {messages.length > 0 &&
        messages.map((message, index) => (
          <CustomToast message={message} key={index} />
        ))}
    </Fragment>
  );
};

export default Home;
