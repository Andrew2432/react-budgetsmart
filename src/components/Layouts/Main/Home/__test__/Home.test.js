import TestRenderer from 'react-test-renderer';
import React from 'react';
import Home from '../Home';
import ToastContext from '../../../../../context/Toasts/ToastContext';
import ExpenseContext from '../../../../../context/Expense/expenseContext';

describe('Home component', () => {
  test('Renders correctly', () => {
    const home = new TestRenderer.create(
      (
        <ExpenseContext.Provider value={{ expenses: [] }}>
          <ToastContext.Provider value={{ messages: [] }}>
            <Home />
          </ToastContext.Provider>
        </ExpenseContext.Provider>
      )
    );
    expect(home).toMatchSnapshot();
  });
});
