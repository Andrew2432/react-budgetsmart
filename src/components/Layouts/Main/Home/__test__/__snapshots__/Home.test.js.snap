// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component Renders correctly 1`] = `
Array [
  <div
    className="card-panel deep-purple white-text"
  >
    <h4>
      Add A New Expense
    </h4>
    <div
      className="row"
    >
      <div
        className="col s8 m8 l8"
      >
        <div
          className="input-field col s12 m12 l12"
        >
          <input
            className=""
            id="expense"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label
            className=""
            htmlFor="expense"
          >
            Expense
          </label>
        </div>
      </div>
      <div
        className="col s4 m4 l4"
      >
        <div
          className="input-field col s12 m12 l12"
        >
          <input
            className=""
            id="cost"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label
            className=""
            htmlFor="cost"
          >
            Cost
          </label>
        </div>
      </div>
    </div>
    <div
      className="row"
    />
  </div>,
  <div
    className="card-panel deep-orange darken-1"
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    <h4>
      Total expense
    </h4>
    <h4>
      Rs. 
    </h4>
  </div>,
  <div
    className="card-panel indigo accent-2"
  >
    <h4>
      Your Expenses
    </h4>
    <ul
      className="collection black-text"
    >
      <li
        className="collection-item"
      >
        No expenses added. Enter an expense
      </li>
    </ul>
  </div>,
]
`;
